quiche/src/lib.rs

```diff
@@ -15633,32 +15790,80 @@ mod tests {
-        let server_scid = ConnectionId::from_ref(&server_scid);
-        let server_addr = "127.0.0.1:4321".parse().unwrap();
+        assert_eq!(pipe.handshake(), Ok(()));
 
-        let mut pipe = testing::Pipe {
-            client: connect(
-                Some("quic.tech"),
-                &client_scid,
-                client_addr,
-                server_addr,
-                &mut client_config,
-            )?,
-            server: accept(
-                &server_scid,
-                None,
-                server_addr,
-                client_addr,
-                &mut server_config,
-            )?,
-        };
+        Ok(())
+    }
+
+    #[cfg(feature = "boringssl-boring-crate")]
+    #[test]
+    fn in_handshake_config() -> Result<()> {
+        let mut buf = [0; 65535];
+
+        // Manually construct `SSlContextBuilder` for the server.
+        let mut server_tls_ctx_builder =
+            boring::ssl::SslContextBuilder::new(boring::ssl::SslMethod::tls())
+                .unwrap();
+        server_tls_ctx_builder
+            .set_certificate_chain_file("examples/cert.crt")
+            .unwrap();
+        server_tls_ctx_builder
+            .set_private_key_file(
+                "examples/cert.key",
+                boring::ssl::SslFiletype::PEM,
+            )
+            .unwrap();
+        server_tls_ctx_builder.set_select_certificate_callback(|mut hello| {
+            Connection::set_initial_congestion_window_packets_in_handshake(
+                hello.ssl_mut(),
+                30,
+            )
+            .unwrap();
+
+            Ok(())
+        });
+
+        let mut server_config = Config::with_boring_ssl_ctx_builder(
+            crate::PROTOCOL_VERSION,
+            server_tls_ctx_builder,
+        )?;
+
+        let mut client_config = Config::new(crate::PROTOCOL_VERSION)?;
+        client_config.load_cert_chain_from_pem_file("examples/cert.crt")?;
+        client_config.load_priv_key_from_pem_file("examples/cert.key")?;
+
+        for config in [&mut client_config, &mut server_config] {
+            config.set_application_protos(&[b"proto1", b"proto2"])?;
+            config.set_initial_max_data(1000000);
+            config.set_initial_max_stream_data_bidi_local(15);
+            config.set_initial_max_stream_data_bidi_remote(15);
+            config.set_initial_max_stream_data_uni(10);
+            config.set_initial_max_streams_bidi(3);
+            config.set_initial_max_streams_uni(3);
+            config.set_max_idle_timeout(180_000);
+            config.verify_peer(false);
+            config.set_ack_delay_exponent(8);
+        }
+
+        let mut pipe = testing::Pipe::with_client_and_server_config(
+            &mut client_config,
+            &mut server_config,
+        )?;
+
+        // Client sends initial flight.
+        let (len, _) = pipe.client.send(&mut buf).unwrap();
+        pipe.server_recv(&mut buf[..len]).unwrap();
+
+        assert_eq!(pipe.server.tx_cap, 36000);
```

Comment:
Consider changing to:

`assert_eq!(pipe.server.tx_cap, icwnd * 1200);`

and using icwnd in the set_select_certificate_callback.

Also, it is possible to query `pipe.server.tx_cap` prior to `pipe.client.send`?  Wondering if we can `assert_eq!(pipe.server.tx_cap, default_icwnd * 1200);`

Reply:
I changed the check to use a constant for the custom initial window, and also added a check for tx_cap before the update, but note that before receiving the client's initial tx_cap is always 0 (as that's the default value for the flow control window).